pkg/runner/test_runner.go

```diff
@@ -645,6 +686,33 @@ func (p *pipeline) waitTargets(terminate <-chan struct{}, completedCh chan<- *ta
 			completedTarget = targetErr.Target
 			completedTargetError = targetErr.Err
 		}
+
+		if err != nil {
+			return err
+		}
+
+		if outChannel == nil {
+			log.Debugf("no more targets to wait, output channel is closed")
+			break
+		}
+
+		if completedTarget != nil {
+			p.state.SetTarget(completedTarget, completedTargetError)
+			log.Debugf("writing target %+v on the completed channel", completedTarget)
+			if err := writer.writeTimeout(terminate, completedCh, completedTarget, p.timeouts.MessageTimeout); err != nil {
+				log.Panicf("could not write completed target: %v", err)
+			}
+			completedTarget = nil
+			completedTargetError = nil
+		}
+
+		numIngress := atomic.LoadUint64(&p.numIngress)
+		numCompleted := uint64(len(p.state.CompletedTargets()))
+		log.Debugf("targets completed: %d, expected: %d", numCompleted, numIngress)
+		if numIngress != 0 && numCompleted == numIngress {
+			log.Debugf("no more targets to wait, all targets (%d) completed", numCompleted)
+			break
+		}
```

Comment:
> Out of curiosity: when it may happened that outChannel is not closed but all targets completed? At first sight (may be I wrong) looks like a light version of a Rube Goldberg machine :), to be specific:

This requires better documentation and diagrams, but the criteria are the following:
* All targets completed is not enough to consider a test completed. The steps should complete as well. Probably there should be a timer, which fires if the steps do not complete after a certain time after seeing all targets completing. At the moment this is not implemented, but I think it's the right thing to do to not terminate until all steps terminate and make sure we don't leak anything behind.

* If `outChannel` is closed, we do run `waitSteps` which waits for routing blocks and steps to complete, with a timeout